File: gen-pass-ts/src/passwordGenerator.ts
import type { RandomUint } from "./random.js";
import { defaultRandomUint } from "./random.js";
import {
  CharClass,
  defaultPasswordConfig,
  GeneratorFlag,
  type PasswordConfig,
} from "./types.js";

export type { PasswordConfig };
export { CharClass, GeneratorFlag, defaultPasswordConfig };

function hasClass(classes: number, flag: number): boolean {
  return (classes & flag) !== 0;
}

function hasFlag(flags: number, flag: number): boolean {
  return (flags & flag) !== 0;
}

export function passwordGroups(config: PasswordConfig): string[][] {
  const exclude = hasFlag(config.flags, GeneratorFlag.EXCLUDE_LOOK_ALIKE);
  const groups: string[][] = [];

  if (hasClass(config.classes, CharClass.LOWER_LETTERS)) {
    const group: string[] = [];
    for (let code = 97; code <= 122; code += 1) {
      if (exclude && code === 108) {
        continue;
      }
      group.push(String.fromCodePoint(code));
    }
    groups.push(group);
  }

  if (hasClass(config.classes, CharClass.UPPER_LETTERS)) {
    const group: string[] = [];
    for (let code = 65; code <= 90; code += 1) {
      if (exclude && (code === 66 || code === 71 || code === 73 || code === 79)) {
        continue;
      }
      group.push(String.fromCodePoint(code));
    }
    groups.push(group);
  }

  if (hasClass(config.classes, CharClass.NUMBERS)) {
    const group: string[] = [];
    for (let code = 48; code <= 57; code += 1) {
      if (exclude && (code === 48 || code === 49 || code === 54 || code === 56)) {
        continue;
      }
      group.push(String.fromCodePoint(code));
    }
    groups.push(group);
  }

  if (hasClass(config.classes, CharClass.BRACES)) {
    groups.push(["(", ")", "[", "]", "{", "}"]);
  }

  if (hasClass(config.classes, CharClass.PUNCTUATION)) {
    groups.push([",", ".", ":", ";"]);
  }

  if (hasClass(config.classes, CharClass.QUOTES)) {
    groups.push(['"', "'"]);
  }

  if (hasClass(config.classes, CharClass.DASHES)) {
    const group = ["-", "/", "\\", "_"];
    if (!exclude) {
      group.push("|");
    }
    groups.push(group);
  }

  if (hasClass(config.classes, CharClass.MATH)) {
    groups.push(["!", "*", "+", "<", "=", ">", "?"]);
  }

  if (hasClass(config.classes, CharClass.LOGOGRAMS)) {
    const group: string[] = [];
    for (let code = 35; code <= 38; code += 1) {
      group.push(String.fromCodePoint(code));
    }
    group.push("@", "^", "`", "~");
    groups.push(group);
  }

  if (hasClass(config.classes, CharClass.EASCII)) {
    const group: string[] = [];
    for (let code = 161; code <= 172; code += 1) {
      group.push(String.fromCodePoint(code));
    }
    for (let code = 174; code <= 255; code += 1) {
      if (exclude && code === 249) {
        continue;
      }
      group.push(String.fromCodePoint(code));
    }
    groups.push(group);
  }

  if (hasClass(config.classes, CharClass.WHITESPACE)) {
    groups.push([" "]);
  }

  if (config.customCharset.length > 0) {
    const group: string[] = [];
    for (const ch of config.customCharset) {
      if (!group.includes(ch)) {
        group.push(ch);
      }
    }
    if (group.length > 0) {
      groups.push(group);
    }
  }

  const excluded = [...config.excludedCharset];
  return groups
    .map((group) => group.filter((ch) => !excluded.includes(ch)))
    .filter((group) => group.length > 0);
}

export function minLength(config: PasswordConfig): number {
  if (hasFlag(config.flags, GeneratorFlag.CHAR_FROM_EVERY_GROUP)) {
    return passwordGroups(config).length;
  }
  return 1;
}

export function isValid(config: PasswordConfig): boolean {
  if (config.classes === 0 && config.customCharset.length === 0) {
    return false;
  }
  if (config.length === 0) {
    return false;
  }
  if (
    hasFlag(config.flags, GeneratorFlag.CHAR_FROM_EVERY_GROUP) &&
    config.length < minLength(config)
  ) {
    return false;
  }
  return passwordGroups(config).length > 0;
}

export function flattenPasswordCharset(config: PasswordConfig): string {
  const chars = passwordGroups(config).flat();
  const unique = [...new Set(chars)].sort();
  return unique.join("");
}

export function generatePasswordWith(
  config: PasswordConfig,
  randomUint: RandomUint,
): string {
  if (!isValid(config)) {
    throw new Error("invalid password generator configuration");
  }

  const groups = passwordGroups(config);
  const passwordChars = groups.flat();
  let password = "";

  if (hasFlag(config.flags, GeneratorFlag.CHAR_FROM_EVERY_GROUP)) {
    for (const group of groups) {
      const pos = randomUint(group.length);
      password += group[pos]!;
    }

    for (let i = groups.length; i < config.length; i += 1) {
      const pos = randomUint(passwordChars.length);
      password += passwordChars[pos]!;
    }

    const chars = [...password];
    for (let i = chars.length - 1; i >= 1; i -= 1) {
      const j = randomUint(i + 1);
      [chars[i], chars[j]] = [chars[j]!, chars[i]!];
    }
    password = chars.join("");
  } else {
    for (let i = 0; i < config.length; i += 1) {
      const pos = randomUint(passwordChars.length);
      password += passwordChars[pos]!;
    }
  }

  return password;
}

export function generatePassword(
  config: PasswordConfig,
  randomUint: RandomUint = defaultRandomUint,
): string {
  return generatePasswordWith(config, randomUint);
}

File: gen-pass-ts/tools/parity-ref/src/password_generator.rs
use bitflags::bitflags;

use crate::random;

bitflags! {
    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    pub struct CharClass: u32 {
        const LOWER_LETTERS = 1 << 0;
        const UPPER_LETTERS = 1 << 1;
        const NUMBERS = 1 << 2;
        const BRACES = 1 << 3;
        const PUNCTUATION = 1 << 4;
        const QUOTES = 1 << 5;
        const DASHES = 1 << 6;
        const MATH = 1 << 7;
        const LOGOGRAMS = 1 << 8;
        const SPECIAL_CHARACTERS = Self::BRACES.bits()
            | Self::PUNCTUATION.bits()
            | Self::QUOTES.bits()
            | Self::DASHES.bits()
            | Self::MATH.bits()
            | Self::LOGOGRAMS.bits();
        const EASCII = 1 << 9;
        const WHITESPACE = 1 << 10;
        const DEFAULT_CHARSET = Self::LOWER_LETTERS.bits()
            | Self::UPPER_LETTERS.bits()
            | Self::NUMBERS.bits();
    }
}

bitflags! {
    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    pub struct GeneratorFlag: u32 {
        const EXCLUDE_LOOK_ALIKE = 1 << 0;
        const CHAR_FROM_EVERY_GROUP = 1 << 1;
        const ADVANCED_MODE = 1 << 2;
        const DEFAULT_FLAGS = Self::EXCLUDE_LOOK_ALIKE.bits() | Self::CHAR_FROM_EVERY_GROUP.bits();
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PasswordGeneratorConfig {
    pub length: u32,
    pub classes: CharClass,
    pub flags: GeneratorFlag,
    pub custom_charset: String,
    pub excluded_charset: String,
}

impl Default for PasswordGeneratorConfig {
    fn default() -> Self {
        Self {
            length: 32,
            classes: CharClass::DEFAULT_CHARSET,
            flags: GeneratorFlag::DEFAULT_FLAGS,
            custom_charset: String::new(),
            excluded_charset: String::new(),
        }
    }
}

type PasswordGroup = Vec<char>;

pub fn password_groups(config: &PasswordGeneratorConfig) -> Vec<PasswordGroup> {
    let exclude = config.flags.contains(GeneratorFlag::EXCLUDE_LOOK_ALIKE);
    let mut groups = Vec::new();

    if config.classes.contains(CharClass::LOWER_LETTERS) {
        let mut group = Vec::new();
        for code in 97..=122 {
            if exclude && code == 108 {
                continue;
            }
            group.push(char::from_u32(code).unwrap());
        }
        groups.push(group);
    }

    if config.classes.contains(CharClass::UPPER_LETTERS) {
        let mut group = Vec::new();
        for code in 65..=90 {
            if exclude && matches!(code, 66 | 71 | 73 | 79) {
                continue;
            }
            group.push(char::from_u32(code).unwrap());
        }
        groups.push(group);
    }

    if config.classes.contains(CharClass::NUMBERS) {
        let mut group = Vec::new();
        for code in 48..=57 {
            if exclude && matches!(code, 48 | 49 | 54 | 56) {
                continue;
            }
            group.push(char::from_u32(code).unwrap());
        }
        groups.push(group);
    }

    if config.classes.contains(CharClass::BRACES) {
        groups.push(vec!['(', ')', '[', ']', '{', '}']);
    }

    if config.classes.contains(CharClass::PUNCTUATION) {
        groups.push(vec![',', '.', ':', ';']);
    }

    if config.classes.contains(CharClass::QUOTES) {
        groups.push(vec!['"', '\'']);
    }

    if config.classes.contains(CharClass::DASHES) {
        let mut group = vec!['-', '/', '\\', '_'];
        if !exclude {
            group.push('|');
        }
        groups.push(group);
    }

    if config.classes.contains(CharClass::MATH) {
        groups.push(vec!['!', '*', '+', '<', '=', '>', '?']);
    }

    if config.classes.contains(CharClass::LOGOGRAMS) {
        let mut group = Vec::new();
        for code in 35..=38 {
            group.push(char::from_u32(code).unwrap());
        }
        group.extend(['@', '^', '`', '~']);
        groups.push(group);
    }

    if config.classes.contains(CharClass::EASCII) {
        let mut group = Vec::new();
        for code in 161..=172 {
            group.push(char::from_u32(code).unwrap());
        }
        for code in 174..=255 {
            if exclude && code == 249 {
                continue;
            }
            group.push(char::from_u32(code).unwrap());
        }
        groups.push(group);
    }

    if config.classes.contains(CharClass::WHITESPACE) {
        groups.push(vec![' ']);
    }

    if !config.custom_charset.is_empty() {
        let mut group = Vec::new();
        for ch in config.custom_charset.chars() {
            if !group.contains(&ch) {
                group.push(ch);
            }
        }
        if !group.is_empty() {
            groups.push(group);
        }
    }

    let excluded: Vec<char> = config.excluded_charset.chars().collect();
    groups.retain_mut(|group| {
        group.retain(|ch| !excluded.contains(ch));
        !group.is_empty()
    });

    groups
}

pub fn min_length(config: &PasswordGeneratorConfig) -> u32 {
    if config.flags.contains(GeneratorFlag::CHAR_FROM_EVERY_GROUP) {
        password_groups(config).len() as u32
    } else {
        1
    }
}

pub fn is_valid(config: &PasswordGeneratorConfig) -> bool {
    if config.classes.is_empty() && config.custom_charset.is_empty() {
        return false;
    }
    if config.length == 0 {
        return false;
    }
    if config.flags.contains(GeneratorFlag::CHAR_FROM_EVERY_GROUP)
        && config.length < min_length(config)
    {
        return false;
    }
    !password_groups(config).is_empty()
}

pub fn generate_password(config: &PasswordGeneratorConfig) -> Result<String, &'static str> {
    generate_password_with(config, |limit| random::random_uint(limit))
}

pub fn generate_password_with(
    config: &PasswordGeneratorConfig,
    mut random_uint: impl FnMut(u32) -> u32,
) -> Result<String, &'static str> {
    if !is_valid(config) {
        return Err("invalid password generator configuration");
    }

    let groups = password_groups(config);
    let mut password_chars = Vec::new();
    for group in &groups {
        password_chars.extend(group.iter().copied());
    }

    let mut password = String::new();

    if config.flags.contains(GeneratorFlag::CHAR_FROM_EVERY_GROUP) {
        for group in &groups {
            let pos = random_uint(group.len() as u32) as usize;
            password.push(group[pos]);
        }

        for _ in groups.len() as u32..config.length {
            let pos = random_uint(password_chars.len() as u32) as usize;
            password.push(password_chars[pos]);
        }

        let mut chars: Vec<char> = password.chars().collect();
        for i in (1..chars.len()).rev() {
            let j = random_uint(i as u32 + 1) as usize;
            chars.swap(i, j);
        }
        password = chars.into_iter().collect();
    } else {
        for _ in 0..config.length {
            let pos = random_uint(password_chars.len() as u32) as usize;
            password.push(password_chars[pos]);
        }
    }

    Ok(password)
}

pub fn flatten_password_charset(config: &PasswordGeneratorConfig) -> String {
    let mut chars: Vec<char> = password_groups(config)
        .iter()
        .flat_map(|group| group.iter().copied())
        .collect();
    chars.sort_unstable();
    chars.dedup();
    chars.into_iter().collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_generates_length_32() {
        let password = generate_password(&PasswordGeneratorConfig::default()).unwrap();
        assert_eq!(password.chars().count(), 32);
    }

    #[test]
    fn char_from_every_group_includes_each_class() {
        let config = PasswordGeneratorConfig {
            length: 8,
            classes: CharClass::LOWER_LETTERS | CharClass::UPPER_LETTERS | CharClass::NUMBERS,
            flags: GeneratorFlag::CHAR_FROM_EVERY_GROUP | GeneratorFlag::EXCLUDE_LOOK_ALIKE,
            ..Default::default()
        };
        let password = generate_password(&config).unwrap();
        assert!(password.chars().any(|c| c.is_ascii_lowercase()));
        assert!(password.chars().any(|c| c.is_ascii_uppercase()));
        assert!(password.chars().any(|c| c.is_ascii_digit()));
    }

    #[test]
    fn exclude_look_alike_chars() {
        let config = PasswordGeneratorConfig {
            length: 64,
            classes: CharClass::LOWER_LETTERS
                | CharClass::UPPER_LETTERS
                | CharClass::NUMBERS
                | CharClass::DASHES,
            flags: GeneratorFlag::CHAR_FROM_EVERY_GROUP | GeneratorFlag::EXCLUDE_LOOK_ALIKE,
            ..Default::default()
        };
        let password = generate_password(&config).unwrap();
        assert!(!password.contains('l'));
        assert!(!password.contains('|'));
        assert!(!password.contains('0'));
    }

    #[test]
    fn excluded_charset_is_removed() {
        let config = PasswordGeneratorConfig {
            length: 16,
            classes: CharClass::LOWER_LETTERS,
            flags: GeneratorFlag::empty(),
            excluded_charset: "aeiou".into(),
            ..Default::default()
        };
        let password = generate_password(&config).unwrap();
        assert!(!password.chars().any(|c| "aeiou".contains(c)));
    }

    #[test]
    fn flatten_password_charset_dedupes_and_excludes() {
        let config = PasswordGeneratorConfig {
            classes: CharClass::LOWER_LETTERS | CharClass::NUMBERS,
            flags: GeneratorFlag::EXCLUDE_LOOK_ALIKE,
            excluded_charset: "0".into(),
            ..Default::default()
        };
        let charset = flatten_password_charset(&config);
        assert!(!charset.contains('0'));
        assert!(!charset.contains('l'));
        assert!(charset.contains('a'));
        assert!(charset.contains('2'));
        let unique: std::collections::HashSet<char> = charset.chars().collect();
        assert_eq!(unique.len(), charset.chars().count());
    }

    #[test]
    fn invalid_config_errors() {
        let config = PasswordGeneratorConfig {
            length: 1,
            classes: CharClass::LOWER_LETTERS | CharClass::UPPER_LETTERS,
            flags: GeneratorFlag::CHAR_FROM_EVERY_GROUP,
            ..Default::default()
        };
        assert!(generate_password(&config).is_err());
    }
}
